src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -410,27 +398,15 @@ else if (null != mHeader) {
      * @param referenceIndex Must either equal NO_ALIGNMENT_REFERENCE_INDEX (-1) indicating no reference, or the
      *                       record must have a SAMFileHeader and the index must exist in the associated sequence
      *                       dictionary.
-     * @throws IllegalStateException if the SAMFileHeader is null for this record or the reference index is not
-     *         found in the sequence dictionary for this record.
+     * @throws IllegalStateException if {@code referenceIndex} is not equal to NO_ALIGNMENT_REFERENCE_INDEX and the
+     * SAMFileHeader is null for this record
+     * @throws IllegalArgumentException if {@code referenceIndex} is not found in the sequence dictionary in the header
+     * for this record.
      */
     public void setReferenceIndex(final int referenceIndex) {
-        if (referenceIndex == NO_ALIGNMENT_REFERENCE_INDEX) {
-            mReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-            mReferenceName = NO_ALIGNMENT_REFERENCE_NAME;
-        }
-        else if (null == mHeader) {
-            throw new IllegalStateException("A non-null SAMFileHeader is required to resolve the reference index");
-        }
-        else {
-            SAMSequenceRecord samSequence = mHeader.getSequence(referenceIndex);
-            if (null != samSequence) {
-                mReferenceIndex = referenceIndex;
-                mReferenceName = samSequence.getSequenceName();
-            }
-            else {
-                throw new IllegalArgumentException("Reference index " + referenceIndex + " not found in sequence dictionary.");
-            }
-        }
+        // resolveNameFromIndex throws if the index can't be resolved
+        setReferenceName(resolveNameFromIndex(referenceIndex, mHeader));
+        mReferenceIndex = referenceIndex;
```

Comment:
Yes, I think I was just trying to be explicit about it since its the main point of the call.